src/Particle/SampleStack.h

```diff
@@ -0,0 +1,91 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2020  QMCPACK developers.
+//
+// File developed by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//
+// File created by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+/** @file SampleStack.h
+ * @brief Stores particle configurations for later use in DMC and wavefunction optimization
+ *
+ * Stores less temporary data than the buffer object.
+ */
+
+#ifndef QMCPLUSPLUS_SAMPLE_STACK_H
+#define QMCPLUSPLUS_SAMPLE_STACK_H
+
+#include <vector>
+#include "Particle/ParticleSet.h"
+#include "Particle/Walker.h"
+
+namespace qmcplusplus
+{
+class MCWalkerConfiguration;
+class HDFWalkerOutput;
+class MCSample;
+
+class SampleStack
+{
+public:
+  typedef ParticleSet::Walker_t Walker_t;
+  typedef std::vector<Walker_t*> WalkerList_t;
+  typedef WalkerList_t::iterator walker_iterator;
+  typedef QMCTraits::PropertySetType PropertySetType;
+
+  SampleStack();
+
+  int setTotalNum(int total_num) { TotalNum = total_num; }
+
+  int getMaxSamples() const { return MaxSamples; }
+
+  void setMaxSamples(int max_samples) { MaxSamples = max_samples; }
+
+  bool empty() const { return sample_stack_.empty(); }
+
+  void putSample(unsigned int i, const Walker_t& w);
+
+  void getSample(unsigned int i, Walker_t& w) const;
```

Comment:
The semantics of this are bad to me why break encapsulation why void with output argument.  This is going to make the refactor I have planned for walker more painful. 
Rather than hide the MCSample type this would make much more sense to me

```C++
MCSample SampleVector::getSample(size_t sample_index)
{
  return MCSample(sample_statck_[sample_index];
}

Walker walker;
walker.loadSample(sample_stack.getSample());
```

Reply:
Some piece of code has to know about both Walker_t and MCSample.   I can change it so that translation code is not here, and expose MCSample outside the class.